flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/schema/DeferredTypeFlinkTableFunction.scala

```diff
@@ -40,6 +43,10 @@ class DeferredTypeFlinkTableFunction(
     val implicitResultType: TypeInformation[_])
   extends FlinkTableFunction(tableFunction, evalMethod) {
 
+  val paramTypeInfos = evalMethod.getParameterTypes.map { paramType =>
+    TypeExtractor.getForClass(paramType)
```

Comment:
I noticed that the semantics of how the type lists are obtained are a bit different in both cases. In `UserDefinedFunctionUtils` we look at the parameters that are handed to the function. Here we look at the signature of the `eval` method. The result will be different if the `eval` method is called with more specific types than its signature has, e.g., if we have `eval(Object o1, Object o2)` and call it with `eval(new Integer(1), new String("Hello"))`. 

Both approaches make sense:
1. parameter types are more specific
2. with signature types it is easier to identify which eval() method is called.

What do you think @clarkyzl, @KurtYoung?